src/inmanta/agent/code_manager.py

```diff
@@ -0,0 +1,127 @@
+"""
+    Copyright 2017 Inmanta
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+    Contact: code@inmanta.com
+"""
+
+import asyncio
+import logging
+import sys
+import uuid
+from typing import Collection
+
+from inmanta import protocol
+from inmanta.agent import executor
+from inmanta.agent.executor import ResourceInstallSpec
+from inmanta.data.model import LEGACY_PIP_DEFAULT, PipConfig, ResourceType
+from inmanta.loader import ModuleSource
+from inmanta.protocol import Client, SyncClient
+from inmanta.util.async_lru import async_lru_cache
+
+LOGGER = logging.getLogger(__name__)
+
+
+class CouldNotResolveCode(Exception):
+
+    def __init__(self, resource_type: str, version: int, error_message: str) -> None:
+        self.msg = f"Failed to get source code for {resource_type} version={version}, result={error_message}"
+        super().__init__(self.msg)
+
+
+class CodeManager:
+    """
+    Helper responsible for translating resource versions into code
+
+    Caches heavily
+    """
+
+    def __init__(self, client: Client) -> None:
+        self._client = client
+
+    @async_lru_cache(maxsize=5)
+    async def get_pip_config(self, environment: uuid.UUID, version: int) -> PipConfig:
+        response = await self._client.get_pip_config(tid=environment, version=version)
+        if response.code != 200:
+            raise Exception("Could not get pip config from server " + str(response.result))
+        assert response.result is not None  # mypy
+        pip_config = response.result["data"]
+        if pip_config is None:
+            return LEGACY_PIP_DEFAULT
+        return PipConfig(**pip_config)
+
+    @async_lru_cache(maxsize=1024)
+    async def get_code_for_type(self, environment: uuid.UUID, version: int, resource_type: ResourceType) -> ResourceInstallSpec:
```

Comment:
Now I look at this, failures are cached as well, that may not be the best idea...